test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs

```diff
@@ -584,6 +523,360 @@ public void GetApiDescription_IncludesResponseFormats_FilteredByType()
+
+            // Assert
+            var description = Assert.Single(descriptions);
+            Assert.Equal(4, description.ParameterDescriptions.Count);
+
+            var id = Assert.Single(description.ParameterDescriptions, p => p.Name == "Id");
+            Assert.Same(ApiParameterSource.Path, id.Source);
+            Assert.Equal(typeof(int), id.Type);
+
+            var quantity = Assert.Single(description.ParameterDescriptions, p => p.Name == "Quantity");
+            Assert.Same(ApiParameterSource.ModelBinding, quantity.Source);
+            Assert.Equal(typeof(int), quantity.Type);
+
+            var productId = Assert.Single(description.ParameterDescriptions, p => p.Name == "Product.Id");
+            Assert.Same(ApiParameterSource.ModelBinding, productId.Source);
+            Assert.Equal(typeof(int), productId.Type);
+
+            var price = Assert.Single(description.ParameterDescriptions, p => p.Name == "Product.Price");
+            Assert.Same(ApiParameterSource.Query, price.Source);
+            Assert.Equal(typeof(decimal), price.Type);
+        }
+
+        // The method under test uses an attribute on the parameter to set a 'default' source
+        [Fact]
+        public void GetApiDescription_ParameterDescription_ComplexDTO_AnotherLevel_AmbientValueProviderMetadata()
+        {
+            // Arrange
+            var action = CreateActionDescriptor(nameof(AcceptsOrderDTO_Query));
+            var parameterDescriptor = action.Parameters.Single();
+
+            // Act
+            var descriptions = GetApiDescriptions(action);
+
+            // Assert
+            var description = Assert.Single(descriptions);
+            Assert.Equal(3, description.ParameterDescriptions.Count);
+
+            var id = Assert.Single(description.ParameterDescriptions, p => p.Name == "Id");
+            Assert.Same(ApiParameterSource.Path, id.Source);
+            Assert.Equal(typeof(int), id.Type);
+
+            var quantity = Assert.Single(description.ParameterDescriptions, p => p.Name == "Quantity");
+            Assert.Same(ApiParameterSource.Query, quantity.Source);
+            Assert.Equal(typeof(int), quantity.Type);
+
+            var product = Assert.Single(description.ParameterDescriptions, p => p.Name == "Product");
+            Assert.Same(ApiParameterSource.Query, product.Source);
+            Assert.Equal(typeof(OrderProductDTO), product.Type);
+        }
+
+        [Fact]
+        public void GetApiDescription_ParameterDescription_BreaksCycles()
+        {
+            // Arrange
+            var action = CreateActionDescriptor(nameof(AcceptsCycle));
+            var parameterDescriptor = action.Parameters.Single();
+
+            // Act
+            var descriptions = GetApiDescriptions(action);
+
+            // Assert
+            var description = Assert.Single(descriptions);
+            Assert.Equal(1, description.ParameterDescriptions.Count);
+
+            var c = Assert.Single(description.ParameterDescriptions);
```

Comment:
yeah i realized that, probably rename the parameter/property and have the same name here ? 
for example ```AcceptsHasCollection(HasCollection c)``` to ```AcceptsHasCollection(HasCollection collection)``` ...
consider this as a nit.